python/lsst/cp/pipe/cpSpectroFlat.py

```diff
@@ -0,0 +1,324 @@
+        inputMD = inputs['inputFlat'].getMetadata()
+        for inputName in sorted(list(inputs.keys())):
+            reference = getattr(inputRefs, inputName, None)
+            if reference is not None and hasattr(reference, "run"):
+                runKey = f"LSST CALIB RUN {inputName.upper()}"
+                runValue = reference.run
+                idKey = f"LSST CALIB UUID {inputName.upper()}"
+                idValue = str(reference.id)
+
+                inputMD[runKey] = runValue
+                inputMD[idKey] = idValue
+        inputs['inputMD'] = inputMD
+
+        outputs = self.run(**inputs)
+        butlerQC.put(outputs, outputRefs)
+
+    def run(self, dummyExpRef, inputFlat, inputPtc, inputMD):
+        """Create spectro flat from an input flat and PTC.
+
+        Parameters
+        ----------
+        dummyExpRef : `lsst.daf.butler.DeferredDatasetHandle`
+            Exposure used to ensure the correct flat is chosen.
+        inputFlat : `lsst.afw.image.Exposure`
+            Input flat to convert to spectro-flat.
+        inputPtc : `lsst.ip.isr.PhotoTransferCurveDataset`
+            Input PTC containing the gains to apply to spectro-flat.
+        inputMD : `lsst.daf.base.PropertyList`
+            Provenance metadata derived from the butler references.
+
+        Returns
+        -------
+        results : `lsst.pipe.base.Struct`
+            The results struct containing:
+
+            ``outputData``
+                Final modified flat generated from the inputs
+                (`lsst.afw.image.Exposure`).
+        """
+        # Verify input
+        exposureFilter = dummyExpRef.get(component="filter")
+        if exposureFilter.physicalLabel != self.config.inputFlatPhysicalFilter:
+            raise RuntimeError("Did not receive expected physical_filter: "
+                               f"{exposureFilter.physicalLabel} != "
+                               f"{self.config.inputFlatPhysicalFilter}")
+
+        detector = inputFlat.getDetector()
+        statsControl = afwMath.StatisticsControl(self.config.backgroundSigmaClipThreshold,
+                                                 5,
+                                                 0x0)
+
+        # Make gain flat
+        # https://github.com/lsst/atmospec/blob/u/jneveu/doFlat/python/lsst/atmospec/spectroFlat.py#L85
+        # TODO: DM-46080 Ensure cpSpectroFlat can distinguish between
+        # new and old flats.
+        gains = inputPtc.gain
+        scaledGains = {}
+        gainFlat = inputFlat.clone()
+
+        # Todo: Get temperature dependence terms.
+        if self.config.scaleGainsByMean:
+            gainMean = np.mean(list(gains.values()))
+        else:
+            gainMean = 1.0
+
+        for amp in detector:
+            bbox = amp.getBBox()
+            gainFlat[bbox].image.array[:, :] = 1 / (gains[amp.getName()] / gainMean)
+            scaledGains[amp.getName()] = gains[amp.getName()] / gainMean
```

Comment:
It would be nice to move this before `gainFlat` and use `scaledGains` in `gainFlat` directly.